legacy/reqcounter/reqcounter.go

```diff
@@ -19,76 +19,127 @@ package reqcounter
+// the counter should ever reset.
 type RequestCounter struct {
-	mutex    sync.Mutex // Lock to prevent race-conditions with concurrent processes.
-	requests uint64     // Number of HTTP requests since recording started.
-	since    time.Time  // When the current counter began recording requests.
+	Mutex    sync.Mutex    // Lock to prevent race-conditions with concurrent processes.
+	Requests uint64        // Number of HTTP requests since recording started.
+	Since    time.Time     // When the current counter began recording requests.
+	Interval time.Duration // The duration of time between each log.
 }
 
-// log records the number of HTTP requests found and resets the counter.
-func (rc *RequestCounter) log() {
+// increment adds 1 to the request counter, signifying another call to GCR.
+func (rc *RequestCounter) Increment() {
+	rc.Mutex.Lock()
+	atomic.AddUint64(&rc.Requests, 1)
+	rc.Mutex.Unlock()
+}
+
+// Flush records the number of HTTP requests found and resets the counter.
+func (rc *RequestCounter) Flush() {
 	// Hold onto the lock when reading & writing the counter.
-	rc.mutex.Lock()
-	defer rc.mutex.Unlock()
+	rc.Mutex.Lock()
+	defer rc.Mutex.Unlock()
 
 	// Log the number of requests within this measurement window.
-	msg := fmt.Sprintf("Since %s there have been %d requests to GCR.", counter.since.Format("2006-01-02 15:04:05"), rc.requests)
-	logrus.Debug(msg)
+	msg := fmt.Sprintf("From %s to %s [%d min] there have been %d requests to GCR.", rc.Since.Format(loggingFormat), time.Now().Format(loggingFormat), rc.Interval/time.Minute, rc.Requests)
+	Debug(msg)
 
 	// Reset the counter.
-	rc.requests = 0
-	rc.since = time.Now()
+	rc.reset()
+}
+
+// reset clears the request counter and stamps the current time of reset.
+// TODO: @tylerferrara in the future, use the request counter field 'persistent' to check
+// if the counter should be reset.
+func (rc *RequestCounter) reset() {
+	rc.Requests = 0
+	rc.Since = time.Now()
+}
+
+// watch continuously logs the counter at the specified intervals.
+func (rc *RequestCounter) watch() {
+	go func() {
+		for {
+			time.Sleep(rc.Interval)
+			rc.Flush()
+		}
+	}()
+}
+
+// NetworkMonitor is the primary means of monitoring network traffic between CIP and GCR.
+type NetworkMonitor struct {
+	Counters []*RequestCounter
+}
+
+// increment adds 1 to each request counter, signifying a new request has been made to GCR.
+func (nm *NetworkMonitor) increment() {
+	// Increment each counter.
+	for _, rc := range nm.Counters {
+		rc.Increment()
+	}
+}
+
+// Log begins logging each counter at their specified intervals.
+func (nm *NetworkMonitor) Log() {
+	for _, rc := range nm.Counters {
+		rc.watch()
+	}
 }
 
 const (
-	// measurementWindow specifies the length of time to wait before logging the RequestCounter. Since Google's
+	// MeasurementWindow specifies the length of time to wait before logging the RequestCounter. Since Google's
```

Comment:
Please add a clarification here that we are not measuring the same 10 minute window that GCR may be measuring. It's a rough approximation.

It may be that GCR measures a sliding 10 minute window every minute. The docs at https://cloud.google.com/container-registry/quotas don't say either way, so we have to live with this ambiguity. Because of this ambiguity, we have to call it out in our own code.

You could also reach out to the GCR team to clear this up (they work at Google ;)

Reply:
Ack.